### PR TITLE
Setup Benchmarks

### PR DESCRIPTION
**Pull request by neekolas**
## Summary

I wanted to setup some basic benchmarking for the JS SDK. 

### What is this useful for

Primarily, these benchmarks are going to be useful for comparing the JS SDK to itself. Does a particular change improve or degrade performance? The performance numbers in the benchmark are not going to be representative of real users, who are using XMTP on devices with very different capabilities. It may be useful for getting a ballpark approximation of performance differences between different SDKs (JS vs iOS), but mobile SDKs are meant to be run on very different devices so the "apples to apples" comparison isn't likely very useful.

## What benchmarks are included

- Message encoding/encryption at different message sizes. 
- Message decoding/decryption at different message sizes.

I'm sure we can all imagine dozens of different benchmarks we might be interested in. You should feel free to add more benchmarks to this package. Some more granular benchmarks seem immediately useful to me looking at these results (what about just encryption separate from encoding?).

## How to run the benchmarks myself

```sh
npm run bench
```

### Encode
CleanShot 2023-02-14 at 08 41 33@2x

```json
{
  "name": "encode",
  "date": "2023-02-14T16:31:34.838Z",
  "version": null,
  "results": [
    {
      "name": "encode and encrypt a 128 byte v1 message",
      "ops": 112,
      "margin": 1.47,
      "percentSlower": 72.82
    },
    {
      "name": "encode and encrypt a 65536 byte v1 message",
      "ops": 95,
      "margin": 0.72,
      "percentSlower": 76.94
    },
    {
      "name": "encode and encrypt a 524288 byte v1 message",
      "ops": 42,
      "margin": 1.55,
      "percentSlower": 89.81
    },
    {
      "name": "encode and encrypt a 128 byte v2 message",
      "ops": 412,
      "margin": 1.19,
      "percentSlower": 0
    },
    {
      "name": "encode and encrypt a 65536 byte v2 message",
      "ops": 216,
      "margin": 1.9,
      "percentSlower": 47.57
    },
    {
      "name": "encode and encrypt a 524288 byte v2 message",
      "ops": 52,
      "margin": 1.3,
      "percentSlower": 87.38
    }
  ],
  "fastest": {
    "name": "encode and encrypt a 128 byte v2 message",
    "index": 3
  },
  "slowest": {
    "name": "encode and encrypt a 524288 byte v1 message",
    "index": 2
  }
}
```

### Decode
CleanShot 2023-02-14 at 08 41 59@2x

```json
{
  "name": "decode",
  "date": "2023-02-14T16:32:10.290Z",
  "version": null,
  "results": [
    {
      "name": "decode and decrypt a 128 byte v1 message",
      "ops": 77,
      "margin": 1.11,
      "percentSlower": 74.67
    },
    {
      "name": "decode and decrypt a 65536 byte v1 message",
      "ops": 62,
      "margin": 1.25,
      "percentSlower": 79.61
    },
    {
      "name": "decode and decrypt a 524288 byte v1 message",
      "ops": 25,
      "margin": 0.65,
      "percentSlower": 91.78
    },
    {
      "name": "decode and decrypt a 128 byte v2 message",
      "ops": 304,
      "margin": 0.75,
      "percentSlower": 0
    },
    {
      "name": "decode and decrypt a 65536 byte v2 message",
      "ops": 149,
      "margin": 1.19,
      "percentSlower": 50.99
    },
    {
      "name": "decode and decrypt a 524288 byte v2 message",
      "ops": 31,
      "margin": 0.95,
      "percentSlower": 89.8
    }
  ],
  "fastest": {
    "name": "decode and decrypt a 128 byte v2 message",
    "index": 3
  },
  "slowest": {
    "name": "decode and decrypt a 524288 byte v1 message",
    "index": 2
  }
}
```

## What's next?
- I'd like to see these benchmarks run inside a real browser instance (likely Puppeteer) to see how that impacts performance
- More benchmarks in the suite
- Benchmarks for more SDKs
- Run benchmarking in CI

### CODE REVIEW

File: src/conversations/Conversation.ts [hunk 1]
```diff
@@ -319,7 +319,7 @@ export class ConversationV2 {
     return this.client.address
   }
 
-  private async encodeMessage(
+  async encodeMessage(
```
Comment (neekolas) [inline src/conversations/Conversation.ts:322]: Needed to expose this in order to use it in benchmarks

File: bench/decode.ts [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+  wrapSuite,
+} from './helpers'
+import { add } from 'benny'
+import { fetcher } from '@xmtp/proto'
+
+const decodeV1 = () => {
+  return MESSAGE_SIZES.map((size) =>
+    add(`decode and decrypt a ${size} byte v1 message`, async () => {
+      const alice = await newLocalHostClient()
+      const bob = await newPrivateKeyBundle()
```
Comment (michaelx11) [inline bench/decode.ts:21]: The benchmark includes setup cost such as creating a new private key bundle, etc right?

**Review by michaelx11 — APPROVED** [review]
awesome stuff!

File: bench/decode.ts (hunk 2 above)
Comment (neekolas) [reply bench/decode.ts:21]: No. The setup happens before the benchmark timer begins. Only the returned function is part of the actual benchmark.

**Review by mkobetic — APPROVED** [review]
Neat. 👍

File: bench/decode.ts (hunk 2 above)
Comment (michaelx11) [reply bench/decode.ts:21]: oh neat!

File: bench/encode.ts [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+  wrapSuite,
+} from './helpers'
+
+const encodeV1 = () => {
+  return MESSAGE_SIZES.map((size) =>
+    add(`encode and encrypt a ${size} byte v1 message`, async () => {
+      const alice = await Client.create(newWallet(), { env: 'local' })
+      const bobKeys = (await newPrivateKeyBundle()).getPublicKeyBundle()
+
+      const message = randomBytes(size)
```
[the remark this answers is not in the document]
Comment (neekolas) [reply bench/encode.ts:20]: At least right now, I don't think there is anywhere that is caching or memoizing anything. But we could pre-create the fixtures and get it to use a different one each run.

**Comment by neekolas** [comment]
> Noice! This is super clean. Out of curiosity, is there a tradeoff to having the result as ops/sec, as opposed to ms/op or ms/100 ops?

That's just the default format of `benny`. We could write a custom hook to transform the output to another format if we felt strongly.

**Pull request merged**